### PR TITLE
[WIP] Custom Inventories

### PR DESCRIPTION
**Pull request by Faithcaio**
[API] (https://github.com/SpongePowered/SpongeAPI/pull/1258) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/752)
___

!! Current Code is now in main repository under ```feature/custom-inventory``` !! so don't bother commenting on code here.

Custom Inventories!

### CODE REVIEW

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+import net.minecraft.inventory.InventoryBasic;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.IChatComponent;
+import net.minecraft.world.IInteractionObject;
+import org.spongepowered.api.item.inventory.custom.InventoryType;
+
+import static org.spongepowered.api.item.inventory.custom.InventoryTypes.*;
+
+public class SpongeCustomInventory implements IInteractionObject, IInventory
+{
```
Comment (me4502) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: These should all be on the previous line

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.item.inventory.custom;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.inventory.*;
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: No star imports please. We've got a formatting file in API's repo under extra.

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.IChatComponent;
+import net.minecraft.world.IInteractionObject;
+import org.spongepowered.api.item.inventory.custom.InventoryType;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TranslatableText;
+import org.spongepowered.api.text.translation.Translation;
+import org.spongepowered.common.text.SpongeTexts;
+
+import static org.spongepowered.api.item.inventory.custom.InventoryTypes.*;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: Don't star static import fields like this.

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+public class SpongeCustomInventory implements IInteractionObject, IInventory {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: Missing new line after this line.

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public int getFieldCount() {
+        return inv.getFieldCount();
+    }
+
+    @Override
+    public void clear() {
+        inv.clear();
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        @Override
+        public boolean canInteractWith(EntityPlayer playerIn) {
+            return true;
+        }
+    }
+
+    public InventoryType getType() {
+        return type;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+import static org.spongepowered.api.item.inventory.custom.InventoryTypes.*;
+
+public class SpongeCustomInventory implements IInteractionObject, IInventory {
+    private InventoryType type;
+    private InventoryBasic inv;
+    private Text displayName;
+
+    public SpongeCustomInventory(Text title, int slotCount, InventoryType type) {
+        boolean isTranslation = title instanceof Translation;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: A `Text` cannot be a `Translation`.

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        // TODO no negative size
+        // TODO default name
+        return (CustomInventory) new SpongeCustomInventory(name, size, type);
+    }
+
+    @Override
+    public Builder type(InventoryType type) {
+        this.type = type;
+        if (type == CHEST) {
+            if (!(size % 9 == 0 && size <= 9 * 6 && size > 0)) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    public CustomInventory build() {
+        // TODO no negative size
+        // TODO default name
+        return (CustomInventory) new SpongeCustomInventory(name, size, type);
+    }
+
+    @Override
+    public Builder type(InventoryType type) {
+        this.type = type;
+        if (type == CHEST) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java]: Should we be changing the size here? In my opinion, we should let the size be set through the builder, and do verification when building the actual inventory, not here.

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 10]
```diff
@@ -24,10 +24,6 @@
  */
 package org.spongepowered.common.mixin.core.entity.player;
 
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Preconditions.checkNotNull;
-import static org.spongepowered.common.entity.CombatHelper.getNewTracker;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: No.

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 11]
```diff
@@ -118,15 +115,10 @@
...
 import org.spongepowered.common.event.CauseTracker;
 import org.spongepowered.common.event.SpongeCommonEventFactory;
-import org.spongepowered.common.interfaces.IMixinCommandSender;
-import org.spongepowered.common.interfaces.IMixinCommandSource;
-import org.spongepowered.common.interfaces.IMixinEntityPlayerMP;
-import org.spongepowered.common.interfaces.IMixinPacketResourcePackSend;
-import org.spongepowered.common.interfaces.IMixinServerScoreboard;
-import org.spongepowered.common.interfaces.IMixinSubject;
-import org.spongepowered.common.interfaces.IMixinTeam;
+import org.spongepowered.common.interfaces.*;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: No...

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    @Override
+    public Container createContainer(InventoryPlayer inventory, EntityPlayer player) {
+        return this.containerProvider.get(this, inventory, player);
+    }
+
+    @Override
+    public String getGuiID() {
+        return guiID;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `this.`

Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `guiId` (lowercase `D`)

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+
+    public IChatComponent getDisplayName() {
+        return SpongeTexts.toComponent(displayName);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        return this.inv.getFieldCount();
+    }
+
+    @Override
+    public void clear() {
+        this.inv.clear();
+    }
+
+    public SpongeCustomInventoryBuilder.CustomContainerProvider getContainerProvider() {
+        return containerProvider;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventory.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    public B size(int size) {
+        this.size = size;
+        return (B) this;
+    }
+
+    @Override
+    public B reset() {
+        name = null;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/SpongeCustomInventoryBuilder.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/custom/MixinCustomGridInventory.java [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    @Override
+    public SlotProvider<IInventory, ItemStack> getSlotProvider() {
+        return this.slots;
+    }
+
+    @Override
+    public Lens<IInventory, ItemStack> getRootLens() {
+        return lens;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/custom/MixinCustomGridInventory.java]: `this.`

**Comment by Faithcaio** [comment]
b00315c559ecaf232fc5d64568f210126e02456a

**Pull request closed without merging**